jax/experimental/sparse/bcoo.py

```diff
@@ -610,6 +614,152 @@ def _bcoo_dot_general_abstract_eval(lhs_data, lhs_indices, rhs, *, dimension_num
 
   return core.ShapedArray(out_shape, lhs_data.dtype)
 
+def _bcoo_correct_oob_indices_impl(data, indices, shape):
+  """Replaces out-of-bound indices in a padded representation with zeros.
+
+  The BCOO format at times must pad indices to indicate unused entries, and it does this
+  with out-of-bounds indices. This can cause problems when passing the data to other
+  backends (like cuSparse) that do not handle this natively. This function converts
+  out-of-bounds indices to zeros, and also zeros-out associated values in the data array.
+
+  Examples:
+
+  >>> M = jnp.array([[1, 0, 2, 3, 0], [0, 0, 0, 4, 0]])
+  >>> bcoo_mat = BCOO.fromdense(M, nse=7)
+  >>> bcoo_mat
+  BCOO(int32[2, 5], nse=7)
+  >>> data = bcoo_mat.data
+  >>> data
+  DeviceArray([1, 2, 3, 4, 0, 0, 0], dtype=int32)
+  >>> indices = bcoo_mat.indices
+  >>> indices
+  DeviceArray([[0, 0],
+               [0, 2],
+               [0, 3],
+               [1, 3],
+               [2, 5],
+               [2, 5],
+               [2, 5]], dtype=int32)
+
+  After the out-of-bound indices correction:
+
+  >>> _, indices = _bcoo_correct_oob_indices_impl(data, indices, shape=(2, 5))
+  >>> indices
+  DeviceArray([[0, 0],
+               [0, 2],
+               [0, 3],
+               [1, 3],
+               [0, 0],
+               [0, 0],
+               [0, 0]], dtype=int32)
+
+  If the input data contains nonzeros at the corrected indices:
+  >>> data = jnp.arange(1, 8)
+  >>> indices = bcoo_mat.indices
+
+  After the out-of-bound indices correction:
+  >>> data, _ = _bcoo_correct_oob_indices_impl(data, indices, shape=(2, 5))
+
+  >>> data
+  DeviceArray([1, 2, 3, 4, 0, 0, 0], dtype=int32)
+  """
+  # TODO: support batch dimension?
+  props = _validate_bcoo(data, indices, shape)
+  if props.n_batch or props.n_dense:
+    raise NotImplementedError("`n_batch` and `n_dense` not yet supported.")
+  mask = (indices >= jnp.array(shape)).any(-1)
+  return jnp.where(mask, 0, data), jnp.where(mask[:, None], 0, indices)
+
+_bcoo_correct_oob_indices = xla.lower_fun(
+    _bcoo_correct_oob_indices_impl, multiple_results=True, new_style=True)
+
+_bcoo_dot_general_default_translation_rule = xla.lower_fun(
+    _bcoo_dot_general_impl, multiple_results=False, new_style=True)
+
+def _bcoo_dot_general_cuda_translation_rule(
+    ctx, avals_in, avals_out, lhs_data, lhs_indices, rhs, *, dimension_numbers,
+    lhs_shape):
+
+  c = ctx.builder
+
+  (lhs_contract, rhs_contract), _ = dimension_numbers
+  _, lhs_indices_aval, _, = avals_in
+  props = _validate_bcoo_indices(lhs_indices_aval, lhs_shape)
+
+  rhs_ndim = len(c.get_shape(rhs).dimensions())
+  if rhs_ndim == 1:
+    bcoo_dot_general_fn = cusparse.coo_matvec
+  elif rhs_ndim == 2:
+    bcoo_dot_general_fn = cusparse.coo_matmat
+    if rhs_contract[0] == 1:
+      rhs = xops.Transpose(rhs, permutation=[1, 0])
```

Comment:
We should catch the `else` case here explicitly and raise an error... I know we've already checked this in another function, but that's one of those "actions at a distance" that will make this hard to follow when we come back to it in the future.

Reply:
Done.